tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp

```diff
@@ -0,0 +1,174 @@
+//===---------- VariableTypeArray.cpp -------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "sourcekitd/VariableTypeArray.h"
+#include "sourcekitd/CompactArray.h"
+#include "SourceKit/Core/LLVM.h"
+#include "SourceKit/Core/LangSupport.h"
+#include "SourceKit/Support/UIdent.h"
+#include "DictionaryKeys.h"
+
+#include "llvm/Support/MemoryBuffer.h"
+
+using namespace SourceKit;
+using namespace sourcekitd;
+
+namespace {
+struct OutVariableType {
+  unsigned VarOffset;
+  unsigned VarLength;
+  const char *PrintedType;
+  unsigned HasExplicitType;
+};
+
+class VariableTypeReader {
+  // A reader for the `VariableType` fields:
+  //  - Variable offset in the source buffer
+  //  - Variable length in the source buffer
+  //  - The printed type
+  //  - Whether the variable has an explicit type annotation
+  CompactArrayReader<unsigned, unsigned, const char *, unsigned> EntryReader;
```

Comment:
I just looked through the code how `ExpressionTypeReader` handles its strings again and I think unfortunately the output stream + offset approach is really necessary. 

This `const char *` points into `sourcekitd`’s address space, which is not readable by another process unless you are using `sourcektid` in process (which is only a debugging feature). 

The way that `ExpressionTypeReader` handles this, is by transmitting the string with all types in the buffer created by `ExpressionTypeArrayBuilder::createBuffer`. That way all strings will actually be transmitted via XPC and are thus addressable by the client process. The string with all types now has a new address (namely one in the `llvm::MemoryBuffer` returned by `createBuffer`) which gets put into the buffer’s header. That header value can then be read by [`ExpressionTypeReader`](https://github.com/apple/swift/blob/775bff003c1a41a5651e961f68a2244527b06be5/tools/SourceKit/tools/sourcekitd/lib/API/ExpressionTypeArray.cpp#L51) and all types within it can be accessed by their offset.

Sorry for the back and forth regarding this.

Reply:
Thanks for digging into this, that makes sense, I'll make sure to revert those changes.